float32[] input_data
float32[] output_data